src/generator.rs

```diff
@@ -0,0 +1,99 @@
+use crate::{
+    blocks::{
+        AIR, BEDROCK, BROWN_TERRACOTTA, RED_SAND, RED_TERRACOTTA, STONE, TERRACOTTA,
+        YELLOW_TERRACOTTA,
+    },
+    chunk::{Chunk, CHUNK_SIZE},
+};
+use bevy::math::Vec3;
+use noise::{NoiseFn, Perlin};
+
+pub struct Generator {
+    pub seed: u32,
+    perlin: Perlin,
+}
+
+impl Generator {
+    pub fn new(seed: u32) -> Generator {
+        Generator {
+            seed,
+            perlin: Perlin::new(seed),
+        }
+    }
+
+    pub fn generate_chunk(&self, chunk: &mut Chunk) {
+        let chunk_origin = chunk.position * CHUNK_SIZE as f32;
+        if chunk_origin.y < 0.0 {
+            return;
+        }
+
+        for x in -1..=CHUNK_SIZE {
+            for y in -1..=CHUNK_SIZE {
+                for z in -1..=CHUNK_SIZE {
+                    let local_position = Vec3::new(x as f32, y as f32, z as f32);
+                    let block_position = chunk_origin + local_position;
+
+                    let block = self.generate_block(block_position);
+                    chunk.set(x as usize, y as usize, z as usize, block);
+                }
```

Comment:
Since we cast `-1` to `usize` here it will be `usize::MAX`. Adding 1 in `chunk.set` will then overflow. If the overflow is intended use wrapping types: https://doc.rust-lang.org/std/num/struct.Wrapping.html

But I find the logic a bit convoluted to start with `-1` and then add `1`. Why not start with `0` and not add `1`?

Reply:
Hmm, I didn't think about negative values for usize. The problem is that the coordinates -1 and 32 are inside of the neighbor chunk. For the meshing I iterate from 0..CHUNK_SIZE, which I find easier to follow. I could refactor the loops to start with 1 and go to CHUNK_SIZE + 1 but I don't like it very much.

Reply:
Thanks for letting me know about the Wrapping Type.